File: apps/api/app/materials.py
"""The material categories a recovery facility works with.

This list is the contract with apps/web. The same keys appear in
apps/web/lib/materials.ts, and the classification prompt is built from the
descriptions below, so a category is added in both places or in neither.
"""

from __future__ import annotations

from typing import Final

MATERIAL_KEYS: Final[tuple[str, ...]] = (
    "plastic",
    "paper",
    "cardboard",
    "metal",
    "glass",
    "organic",
    "textiles",
    "ewaste",
    "other",
)

MATERIAL_GUIDE: Final[dict[str, str]] = {
    "plastic": (
        "bottles, films, bags, rigid packaging, containers, plastic caps"
    ),
    "paper": "newspaper, office paper, magazines, envelopes, mixed paper",
    "cardboard": "corrugated boxes, cartons, thick board, beverage cartons",
    "metal": "cans, foil, tins, wire, ferrous and non ferrous scrap",
    "glass": "bottles, jars, broken glass pieces",
    "organic": "food waste, garden waste, wood, other wet fraction",
    "textiles": "cloth, rags, fabric offcuts, footwear, carpet",
    "ewaste": (
        "cables, circuit boards, batteries, small electronic appliances"
    ),
    "other": (
        "anything that clearly does not belong to the categories above, "
        "including unidentifiable fragments"
    ),
}


def guide_lines() -> str:
    """The category list as it appears inside the prompt."""
    return "\n".join(
        f"- {key}: {description}" for key, description in MATERIAL_GUIDE.items()
    )


def normalise(value: str) -> str:
    """Map whatever the model returned onto a known key.

    The model is asked for an exact key, but a stray plural, capital letter or
    hyphen should not cost a detection.
    """
    cleaned = value.strip().lower().replace("-", "").replace("_", "")
    cleaned = cleaned.replace(" ", "")

    if cleaned in MATERIAL_KEYS:
        return cleaned

    aliases = {
        "ewaste": "ewaste",
        "electronicwaste": "ewaste",
        "electronics": "ewaste",
        "plastics": "plastic",
        "papers": "paper",
        "metals": "metal",
        "glasses": "glass",
        "textile": "textiles",
        "cloth": "textiles",
        "organics": "organic",
        "food": "organic",
        "carton": "cardboard",
        "corrugated": "cardboard",
    }
    return aliases.get(cleaned, "other")

File: apps/web/lib/materials.ts
/**
 * The material categories a recovery facility works with.
 *
 * This list is the contract between the two applications. apps/api sends these
 * exact keys back in every detection, and the classification prompt is built
 * from the same list, so adding a category here means adding it in
 * apps/api/app/materials.py as well.
 */

export const MATERIAL_KEYS = [
  "plastic",
  "paper",
  "cardboard",
  "metal",
  "glass",
  "organic",
  "textiles",
  "ewaste",
  "other",
] as const;

export type MaterialKey = (typeof MATERIAL_KEYS)[number];

export interface MaterialMeta {
  key: MaterialKey;
  label: string;
  /** Tailwind class fragment resolved from the theme tokens in globals.css. */
  color: string;
  description: string;
}

export const MATERIALS: Record<MaterialKey, MaterialMeta> = {
  plastic: {
    key: "plastic",
    label: "Plastic",
    color: "var(--color-plastic)",
    description: "Bottles, films, rigid packaging and containers.",
  },
  paper: {
    key: "paper",
    label: "Paper",
    color: "var(--color-paper)",
    description: "Newspaper, office paper, magazines and mixed paper.",
  },
  cardboard: {
    key: "cardboard",
    label: "Cardboard",
    color: "var(--color-cardboard)",
    description: "Corrugated boxes, cartons and thick board.",
  },
  metal: {
    key: "metal",
    label: "Metal",
    color: "var(--color-metal)",
    description: "Cans, foil, and ferrous or non ferrous scrap.",
  },
  glass: {
    key: "glass",
    label: "Glass",
    color: "var(--color-glass)",
    description: "Bottles, jars and broken glass pieces.",
  },
  organic: {
    key: "organic",
    label: "Organic",
    color: "var(--color-organic)",
    description: "Food waste, garden waste and other wet fraction.",
  },
  textiles: {
    key: "textiles",
    label: "Textiles",
    color: "var(--color-textiles)",
    description: "Cloth, rags, fabric offcuts and footwear.",
  },
  ewaste: {
    key: "ewaste",
    label: "E-waste",
    color: "var(--color-ewaste)",
    description: "Cables, boards, batteries and small electronics.",
  },
  other: {
    key: "other",
    label: "Other",
    color: "var(--color-other)",
    description: "Anything that does not fall in the categories above.",
  },
};

export const MATERIAL_LIST: MaterialMeta[] = MATERIAL_KEYS.map(
  (key) => MATERIALS[key],
);

export function materialMeta(key: string): MaterialMeta {
  return MATERIALS[key as MaterialKey] ?? MATERIALS.other;
}
